### DISCUSSION TITLE
numdifftools usage in lmfit

### DISCUSSION THREAD

**Discussion by eendebakpt** [lmfit/lmfit-py#806] [Q&A] [answered]
The numdifftools package is used in lmfit for calculation of the covariance matrix. The package is giving various deprecation warnings on our platform. I created a [PR](https://github.com/pbrod/numdifftools/pull/59) to address one of the warnings, but the numdifftools author is not responding.
I anyone from the lmfit community in contact with the numdifftools author? Or are there alternative options we could use in lmfit?

**Comment by reneeotten** [comment, marked as answer]
I added the `numdifftools` calculation of the covariance matrix to `lmfit`, but have no contact with the author. It appears that the NumPy deprecation warnings were introduced in version 1.22.0 which means there is still a fair amount of time for the `numdifftools` author to resolve this before it becomes a real issue.

If the warnings really bother you, you can for sure just suppress them - I don't recall the exact syntax but Google will know ;) I have no idea if there are alternatives, but have certainly no intentions to change anything in `lmfit` until the `numdifftools` package doesn't work anymore because of NumPy failures. But again, presumably `numdifftools` will have resolved it by then.

**Comment by newville** [comment]
@eendebakpt @reneeotten  Yeah, I see those deprecations too when running our tests.  I agree that a) they are really deprecations that `numdifftools` should resolve - and thanks very much for that PR, and b) we don't really want to have to re-tool just because releases of `numdifftools` is a little slower to fix deprecations than we might like. 
   
And, maybe related: we should probably plan to push 1.0.4 relatively soon to avoid similar views on slow progress!

I don't know that author myself, but would be willing to write to them if we thought that would be helpful.

**Reply by peendebak** [reply]
The warnings have been addressed in https://github.com/pbrod/numdifftools/pull/63

**Comment by newville** [comment]
@eendebakpt Thanks!  Since these are being addressed upstream,  don't think there is any need to do anything more about the warnings on our end.

**Comment by neoncloud** [comment]
Speaking of `numdifftools` alternatives, has anyone tried an automatic differential engine? Torch may be too heavy, but there are some lightweight autograd engines that work well with numpy, such as [tinygrad](https://github.com/geohot/tinygrad) and [Autograd](https://github.com/HIPS/autograd).

**Reply by reneeotten** [reply]
please don't hijack old topics for a new discussion, instead open a new one ;) What would be the benefit of changing to any of them compared to the use of `numdifftools`? The only thing we want from the package is a Hessian matrix.
